Bookmarks caching and performance optimizations

symbolist opened the pull request:
The purpose of this PR is to ensure that bookmark titles and breadcrumbs stay consistent with the state of the XBlock in mongo.

One known issue is that if a block has multiple parents, and a student does not have access to all of them, `get_path` may return an ancestor path that they would never see. But currently the behaviour in other places is not defined either e.g. `block.get_parent()` picks one single parent out of all and returns it; which one it picks depends on the modulestore. It may be possible to fetch all ancestor paths and pick one depending on the response to `has_access`.

symbolist commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -60,6 +73,9 @@ def get_path(block):
         """
+
+        # This method may not return the "correct" result if one block has multiple parents, and the
+        # student does not have access to one of those parents.
```
We may be able to fix this.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -21,16 +22,24 @@ class Bookmark(models.Model):
+    block_cache = models.ForeignKey(XBlockCache, null=True, blank=True)
+
+    @property(self):
+        return self.block_cache.display_name
```
Wrong syntax. @property needs to wrap a function.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -21,16 +22,24 @@ class Bookmark(models.Model):
         """
+        if self.updated < self.block_cache.updated:
+            block = modulestore().get_item(self.usage_key)
+            self.path = cls.get_path(block)
```
`cls` is undefined here.

cpennington replied:
Also, there is already a `path_to_location` method that may do what you need.

cpennington replied:
(in `modulestore/search.py`)

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -52,6 +61,10 @@ def create(cls, bookmarks_data, block=None):
+        bookmark.block_cache = XBlockCache.objects.get_or_create(
+            course_key=bookmark.course_key, usage_key=usage_key,
+        )
+        bookmark.save()
```
`get_or_create` returns a tuple of `(object, created)`. You don't need to save if `created` is true. Also, you should be passing in `defaults={...}` to set attributes in the case that the XBlockCache doesn't yet exist.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -69,3 +85,26 @@ def get_path(block):
+        """
+        Set data as JSON on the _paths field.
+        """
+        self._path = json.dumps(value)
```
I'm up in the air about whether it would be better to create (or find an existing) `JsonBlob` field type, rather than having an `_paths` field, and then a `paths` property. The latter version isn't going to play well with all of the standard django features.

cpennington commented on lms/djangoapps/bookmarks/signals.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@receiver(SignalHandler.course_published)
+def listen_for_course_publish(sender, course_key, **kwargs):  # pylint: disable=unused-argument
```
I would name this based on its action, rather than it's condition (which is implied by the `@reciever` decorator).

cpennington commented on lms/djangoapps/bookmarks/signals.py:
```diff
@@ -0,0 +1,13 @@
+@receiver(SignalHandler.course_published)
+def listen_for_course_publish(sender, course_key, **kwargs):  # pylint: disable=unused-argument
+    # Import tasks here to avoid a circular import.
+    from .tasks import update_xblocks_cache
```
There is no circular import with `tasks.py`.

cpennington commented on lms/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,80 @@
+    Updates the XBlocks cache for a course.
+    """
+    # Import here to avoid circular import.
+    from .models import XBlockCache
```
There is no circular import.

cpennington commented on lms/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,80 @@
+    from .models import XBlockCache
+
+    # Ideally we'd like to accept a CourseLocator; however, CourseLocator is not JSON-serializable (by default) so
+    # Celery's delayed tasks fail to start. For this reason, callers should pass the course key as a Unicode string.
```
It would be worth looking to see if there's some way that we can tell celery about this custom data type, so that everyone can simply pass `OpaqueKey` objects to tasks.

cpennington commented on lms/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,80 @@
+        if block_cache.display_name != block_data["display_name"] || block_cache.paths != block_data["paths"]:
+            block_cache.display_name = block_data["display_name"]
+            block_cache.paths = block_data["paths"]
+            block_cache.save()
```
Again, `get_or_create` returns a tuple, and can be used to create the object w/o an additional call to save. Sadly, it looks like you can't just do `update_or_create` until django 1.7 (https://docs.djangoproject.com/en/1.7/ref/models/querysets/#update-or-create)

cpennington commented on lms/djangoapps/bookmarks/views.py:
```diff
@@ -107,7 +107,7 @@ def get_queryset(self):
             return []
 
-        results_queryset = Bookmark.objects.filter(course_key=course_key, user=self.request.user).order_by('-created')
+        results_queryset = Bookmark.objects.filter(course_key=course_key, user=self.request.user).select_related('block_cache').order_by('-created')
```
Please add a test asserting the number of queries needed to load bookmarks (and also the number of queries used when creating new bookmarks or updating existing bookmarks).

symbolist commented:
Hi @ormsbee, can you give this a look? Right now the code is a bit rough (so not ready for code review), but I am interested in whether you think this design makes sense and would be acceptably performant? Also do you have any thoughts on using a JsonBlob field type?

symbolist commented:
@muzaffaryousaf Can you give this a review?

symbolist commented:
@cpennington I have cleaned this up. Will you be able to review it in detail?

Main highlights:
1. Adds the XBlockCache class which keeps track of block display names and paths. Is it okay to keep this model in the bookmarks app?
2. The BookmarksService caches bookmarks for a (user,course) on first read call so that subsequent calls in the request can be served without making database queries. The main use case right now is that the sequence will need to know the bookmark status of all its children and calls `is_bookmarked` multiple times.
3. Also adds `select_related` to existing queries. With this all read and delete calls to the api require one sql query. I have added query count asserts in the tests as well.

Thanks.

symbolist commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -60,12 +123,63 @@ def get_path(block):
+        for ancestor_usage_key in path:
+            if ancestor_usage_key != usage_key and ancestor_usage_key.block_type != 'course':
+                try:
+                    block = modulestore().get_item(ancestor_usage_key)
```
@cpennington I have tweaked `path_to_location` to return the whole path. However, it returns `UsageKeys` and not blocks. So they need to be fetched here. Is there a way to get them all in one query?

cpennington replied:
`modulestore.get_items(<list of usage keys>)`

symbolist replied:
@cpennington I am unable to `get_items()` to accept a list of UsageKeys like this. And the docstring doesn't say anything about this usage either: https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/__init__.py#L761

cpennington replied:
Oh, you're quite correct. I misremembered what the api for that function was. For the moment, it looks like you'll just have to call `get_item` on each block individually.

cpennington commented on common/lib/xmodule/xmodule/modulestore/search.py:
```diff
@@ -6,7 +6,7 @@
 
 
-def path_to_location(modulestore, usage_key):
+def path_to_location(modulestore, usage_key, full_path=False):
```
Optional: This might be cleaner as a pair of methods, one of which finds the full path to the root, and the other of which parses that into `(course_id, chapter, section, position)`.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -36,23 +49,73 @@ def create(cls, bookmark_data):
             ItemNotFoundError: If no block exists for the usage_key.
         """
-        usage_key = bookmark_data.pop('usage_key')
+        data = copy.deepcopy(data)
```
I would advise against `deepcopy` in general. It can be very performance intensive. Looking at the code below, I think you only need a shallow copy to protect the caller from the mutation that you're doing (`dict(data)`), or you could just write in the API that this method may mutate the argument.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -60,12 +123,63 @@ def get_path(block):
         Returns:
             list of dicts of the form {'usage_id': <usage_id>, 'display_name': <display_name>}.
```
Please make these named tuples. Generally speaking, if you have a dict with a fixed set of keys, it's better to use a named tuple so that no one accidentally mutates the dictionary, and so that you could upgrade to a class in the future.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -36,23 +49,73 @@ def create(cls, bookmark_data):
+        """
+        if self.modified < self.xblock_cache.modified:
+
+            if self.xblock_cache.paths and len(self.xblock_cache.paths) == 1:
```
Is the length check attempting to guard against an empty list of paths in `self.xblock_cache.paths`? Or against both that and also a list of paths that's longer than 1? If the former, than the first clause suffices, because `bool([]) == False`. If the latter, why not just pick pick the first cached path (in the case that there are more than one), rather than recomputing all of the paths again?

symbolist replied:
@cpennington This check if for both. The reason I am not picking the first one here is that it may be different from what `jump_to_id` will take the user to, and that will be consistent with what `path_to_location` returns.

cpennington replied:
Is that because `xblock_cache.paths` is not ordered per-user?

symbolist replied:
Nope, `xblock_cache.paths` is the list of all possible paths to the block. Its calculated by doing a DFS of the course tree without reference to any user.

On the other hand the path that `jump_to_id` and `path_to_location` construct is by doing `get_parent()`. This is also done without reference to any user and in case of DAGS, as far as I understand which path is returned depends on the modulestore.

So there can be two cases:
1. The DFS sets only one path in `xblock_cache.paths`. So this will also be the path `jump_to_id` and `path_to_location` return and we can just use it without any worry.
2. However, if the DFS sets more than one path in `xblock_cache.paths`, we don't know here which one the user will be taken to by `jump_to_id`, which is used by the UI. So we call `path_to_location` to get hold of this.

I'll add all this as comments.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -60,12 +123,63 @@ def get_path(block):
+    usage_key = LocationKeyField(max_length=255, db_index=True, unique=True)
+
+    display_name = models.CharField(max_length=255, default='')
+    paths = JSONField(null=True, blank=True)
```
Is this actually multiple paths? Or just a single path?

symbolist replied:
Multiple paths. I'll add some documentation to explain this.

cpennington commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -60,12 +123,63 @@ def get_path(block):
+        Returns:
+            An XBlockCache object.
+        """
+        data = copy.deepcopy(data)
```
I don't think you need this `deepcopy`.

cpennington commented on lms/djangoapps/bookmarks/services.py:
```diff
@@ -31,7 +59,13 @@ def bookmarks(self, course_key):
-        return api.get_bookmarks(self._user, course_key=course_key, fields=DEFAULT_FIELDS + OPTIONAL_FIELDS)
+        bookmarks = api.get_bookmarks(self._user, course_key=course_key, fields=DEFAULT_FIELDS + OPTIONAL_FIELDS)
+
+        if course_key.run is not None:
```
If `course_key.run is None`, what condition does that indicate? Why can't we cache in that case? What kind of performance losses do you get if you can't cache?

symbolist replied:
I got this issue in tests that in some cases the run was missing, the `course_key` could not be serialized and so the `cache_key` could not be constructed. Is there a way to work around that? I don't suppose this will be an issue for a valid course?

cpennington replied:
Well, you might want to use `fill_in_run` and then check. And if the run can't be filled in, then that means that you don't have a real course, so I think I'd lean towards not returning any bookmarks.

cpennington commented on lms/djangoapps/bookmarks/services.py:
```diff
@@ -43,13 +77,20 @@ def is_bookmarked(self, usage_key):
+            if bookmark['usage_id'] == unicode(usage_key):
+                return True
+
+        return False
```
return `unicode(usage_key) in request_cache.data.get(cache_key, [])`?

cpennington replied:
Oh, no, nevermind.

cpennington commented on lms/djangoapps/bookmarks/services.py:
```diff
@@ -43,13 +77,20 @@ def is_bookmarked(self, usage_key):
+            self.bookmarks(course_key)
+
+        for bookmark in request_cache.data.get(cache_key, []):
+            if bookmark['usage_id'] == unicode(usage_key):
```
Nit: you're going to serialize usage key once per bookmarked item, rather than just once.

cpennington commented on lms/djangoapps/bookmarks/services.py:
```diff
@@ -79,6 +122,8 @@ def unset_bookmarked(self, usage_key):
         Returns:
             Bool indicating whether the bookmark was removed.
         """
+        self._clear_bookmarks_cache(usage_key.course_key)
```
Why wipe the whole bookmark cache, rather than just overwriting the single key that the user is changing?

symbolist replied:
Of course, that sounds faster. I'll update it.

cpennington commented on lms/djangoapps/bookmarks/tests/test_views.py:
```diff
@@ -148,54 +168,84 @@ class BookmarksListViewTests(BookmarksViewTestsMixin):
+        if check_all_fields:
+            query_parameters += '&fields=path,display_name'
+
+        with self.assertNumQueries(7):
```
It's often helpful to list out the 7 expected queries, so that the next time someone comes to this line, they know what queries to look for.

symbolist commented:
@cpennington Besides that one issue with `get_items` that I mentioned above and a few quality issues that I am working on, this is ready for a re-review. Thanks!

cpennington commented:
The only other thing I'd like to see is assertions of mongo query counts when doing bookmark operations (especially, when caching all of the paths for a course, and when computing the single path for a user).

muzaffaryousaf commented:
@symbolist There are plenty of quality issues in this PR, can you run quality ?

muzaffaryousaf commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -8,9 +10,39 @@
+
+def prepare_path_for_serialization(path):
+    """
+    Return the data from a list of PathItems ready for serialization to jsonn.
```
json instead of `jsonn`

muzaffaryousaf commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -19,16 +51,23 @@ class Bookmark(TimeStampedModel):
-    path = JSONField(help_text='Path in course tree to the block')
+    _path = JSONField(db_column='path', help_text='Path in course tree to the block')
+
+    xblock_cache = models.ForeignKey('bookmarks.XBlockCache', null=True)
```
I don't understand why we are using ForeignKey relation here. My understanding is, there will be a single xblock_cache object against any Bookmark so why not OneToOne, correct me if i'm wrong.

muzaffaryousaf commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -19,16 +51,23 @@ class Bookmark(TimeStampedModel):
+
+    xblock_cache = models.ForeignKey('bookmarks.XBlockCache', null=True)
+
+    class Meta:
```
This is old style definition use this instead `class Meta(object):`.

muzaffaryousaf commented on lms/djangoapps/bookmarks/models.py:
```diff
@@ -36,36 +75,165 @@ def create(cls, bookmark_data):
+
+
+class XBlockCache(TimeStampedModel):
+
```
This needs doc-string.

muzaffaryousaf commented on lms/djangoapps/bookmarks/services.py:
```diff
@@ -5,22 +5,55 @@
     """
     A service that provides access to the bookmarks API.
+
+    When is_bookmarked() is called the first time, the service
```
It could be like this `When is_bookmarked() or bookmarks()`

muzaffaryousaf commented on lms/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,120 @@
+        for block_cache in block_caches:
+            block_data = blocks_data.pop(unicode(block_cache.usage_key))
+            paths = _paths_from_data(block_data['paths'])
+            if block_cache.display_name != block_data['display_name'] or block_cache.paths != paths_data:
```
I'm wondering about the `paths_data` maybe you needed to use `paths` instead.

muzaffaryousaf commented on lms/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,120 @@
+                'paths': paths,
+            })
+
+            if not created:
```
Can we factor out this functionality and make a helper method.

symbolist commented:
@muzaffaryousaf I have addressed your comments. I have also moved the app from `lms/djangoapps` to `openedx/core/djangoapps` so that the task can be executed when studio sends `course_publish` signals.

@cpennington I have added assertions of mongo query counts for the two methods which access the modulestore. All those changes are in the last commit. Is this what you were looking for?

symbolist commented on openedx/core/djangoapps/bookmarks/tests/test_models.py:
```diff
@@ -0,0 +1,432 @@
+                        for __ in range(children_per_block):
+                            blocks_at_next_level += [ItemFactory.create(
+                                parent_location=block.scope_ids.usage_id, display_name=unicode(display_name)
+                            )]
```
@cpennington However, I am running into an issue here. When using the split store, multiple calls in the `for` loop to `ItemFactory.create()` return blocks which have the same `usage_id`. Any idea what could be causing this? This creates an incorrect tree.

symbolist commented:
@cpennington I have added `bulk_operations()` in 3 places in the code and looks like things are really optimal now. Will you be able to give it a look?

To get around the bug in splits's `publish()`, I am not using `bulk_operations()` in tests which use split for now. It  slows down the tests but they are at least correct. We need to run load-testing on this api, so will try to figure that issue out in parallel.

But besides this one issue in the tests, is this good to go? Or do you have anything else to add?

Thanks.

symbolist commented on openedx/core/djangoapps/bookmarks/models.py:
```diff
@@ -0,0 +1,255 @@
+        usage_key = data.pop('usage_key')
+        usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
+
+        with modulestore().bulk_operations(usage_key.course_key):
```
Added `bulk_operations()` here.

symbolist commented on openedx/core/djangoapps/bookmarks/models.py:
```diff
@@ -0,0 +1,255 @@
+        Returns:
+            list of PathItems
+        """
+        with modulestore().bulk_operations(usage_key.course_key):
```
Added `bulk_operations()` here.

symbolist commented on openedx/core/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,155 @@
+
+    This data consists of the display_name and path of the block.
+    """
+    with modulestore().bulk_operations(course_key):
```
Added `bulk_operations()` here.

symbolist commented on openedx/core/djangoapps/bookmarks/tests/test_models.py:
```diff
@@ -0,0 +1,469 @@
+        """
+        In case of mongo, 1 query is used to fetch the block, and 2 by path_to_location(), and then
+        1 query per parent in path is needed to fetch the parent blocks.
+        """
```
Mongo queries when calling `Bookmark.create()`.

symbolist commented on openedx/core/djangoapps/bookmarks/tests/test_models.py:
```diff
@@ -0,0 +1,469 @@
+    def test_get_path_queries(self, store_type, children_per_block, depth, expected_mongo_calls):
+        """
+        In case of mongo, 2 queries are used by path_to_location(), and then
+        1 query per parent in path is needed to fetch the parent blocks.
```
Mongo queries just for `Bookmark.get_path()`.

symbolist commented on openedx/core/djangoapps/bookmarks/tests/test_tasks.py:
```diff
@@ -0,0 +1,162 @@
+        (ModuleStoreEnum.Type.split, 4, 2, 3),
+        (ModuleStoreEnum.Type.split, 2, 3, 3),
+        (ModuleStoreEnum.Type.split, 2, 4, 3),
+    )
```
Mongo queries when running the task.

cpennington commented:
I see no other structural issues.

:+1:

symbolist commented:
Jenkins, test this please.

muzaffaryousaf commented on openedx/core/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,155 @@
+    """
+    Construct a list of paths from path data.
+    """
+    from .models import PathItem
```
This import shows some circular dependencies, can we move `PathItem` to `init` to avoid such imports.

muzaffaryousaf commented on openedx/core/djangoapps/bookmarks/tests/test_models.py:
```diff
@@ -0,0 +1,469 @@
+        """
+        self.assertEqual(xblock_cache.usage_key, data['usage_key'])
+        self.assertEqual(xblock_cache.course_key, data['usage_key'].course_key)
+        self.assertEqual(xblock_cache.display_name, data['display_name'])
```
Can we also verify _paths field ?

muzaffaryousaf commented on openedx/core/djangoapps/bookmarks/tests/test_models.py:
```diff
@@ -0,0 +1,469 @@
+    @ddt.unpack
+    @mock.patch('openedx.core.djangoapps.bookmarks.models.Bookmark.get_path')
+    def test_path(self, seconds_delta, paths, get_path_call_count, mock_get_path):
+
```
Doc strings missing.

muzaffaryousaf commented on openedx/core/djangoapps/bookmarks/tests/test_models.py:
```diff
@@ -0,0 +1,469 @@
+            self.assertEqual(len(path), depth - 2)
+
+    def test_get_path_in_case_of_exceptions(self):
+
```
Add doc strings here too.

muzaffaryousaf commented on openedx/core/djangoapps/bookmarks/tasks.py:
```diff
@@ -0,0 +1,155 @@
+
+
+@task(name=u'openedx.core.djangoapps.bookmarks.tasks.update_xblock_cache')
+def update_xblocks_cache(course_id):
```
It would be nice if we can add some logging in this file, as this is a background task and logging would be really helpful.

muzaffaryousaf commented on lms/djangoapps/bookmarks/migrations/0003_auto__add_unique_bookmark_user_usage_key.py:
```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
```
Reminder: we need to move remaining migration files from lms to openedx.

muzaffaryousaf commented:
@symbolist i'm done with another review, things are in better shape now, except some few inline comments; Also, it would be nice if can add bokchoys to test the bookmarks with publishing the changed unit name from studio; to verify whether or not they are affecting the bookmark breadcrumb trail. Once the changes are made and tests are passing :+1:  
Also, squash your commits before merging.
Thanks

muzaffaryousaf commented:
test this please

muzaffaryousaf commented:
test this please
